### PR TITLE
API: harmonize Index.to_flat_index to MultiIndex.to_flat_index

### LINKED ISSUE #23670: Rethinking ``to_flat_index`` on flat ``Index``
With reference to #22866, and to a comment I made ( https://github.com/pandas-dev/pandas/pull/22866#issuecomment-437700886 ) which I thought was paranoid and maybe is not.

Three considerations:
1. ``to_flat_index`` is the only method I can think of (the only operation in pandas, actually) which gives a different result when called on a flat ``Index`` and when called on an equivalent 1-level ``MultiIndex``: you obtain the ``Index`` itself (each item being assumingly a scalar) in the first case and an ``Index`` of length 1 tuples in the second case
2. if one needs, for any reason, an ``Index`` of length-one scalar, there is no simple way to obtain it. Or if you prefer: if we add some ``sep`` or analogous arguments (e.g. ``fmt``) to ``to_flat_index``, then it is not going to work on a flat ``Index`` (at least with the current implementation). In general, ``to_flat_index`` is a method which makes sense also because it can be combined with, for instance ``.map``. When doing so, it is good to know for sure that the arguments received by the callable are always tuples.
3. ``Index.to_flat_index()`` as it is now is idempotent and pretty useless: it makes sense for compatibility... but then compatibility is a priority!

These all lead me to conclude: shouldn't ``pd.Index.to_flat_index()`` return an ``Index`` of length-1 tuples?

@WillAyd 

#### Code Sample, a copy-pastable example if possible

```python
In [2]: pd.Index([1, 2, 3,]).to_flat_index()
Out[2]: Int64Index([1, 2, 3], dtype='int64')

In [3]: pd.MultiIndex.from_arrays([pd.Index([1, 2, 3,])]).to_flat_index()
Out[3]: Index([(1,), (2,), (3,)], dtype='object')
```

#### Expected Output

``Out[2]`` should be equal to ``Out[3]``

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: 454ecfc61891be610fbb586a7d946eb61b87f32d
python: 3.5.3.final.0
python-bits: 64
OS: Linux
OS-release: 4.9.0-8-amd64
machine: x86_64
processor: 
byteorder: little
LC_ALL: None
LANG: it_IT.UTF-8
LOCALE: it_IT.UTF-8

pandas: 0.24.0.dev0+995.g454ecfc61
pytest: 3.5.0
pip: 9.0.1
setuptools: 39.2.0
Cython: 0.28.4
numpy: 1.14.3
scipy: 0.19.0
pyarrow: None
xarray: None
IPython: 6.2.1
sphinx: 1.5.6
patsy: 0.5.0
dateutil: 2.7.3
pytz: 2018.4
blosc: None
bottleneck: 1.2.0dev
tables: 3.3.0
numexpr: 2.6.1
feather: 0.3.1
matplotlib: 2.2.2.post1634.dev0+ge8120cf6d
openpyxl: 2.3.0
xlrd: 1.0.0
xlwt: 1.3.0
xlsxwriter: 0.9.6
lxml: 4.1.1
bs4: 4.5.3
html5lib: 0.999999999
sqlalchemy: 1.0.15
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: 0.2.1
gcsfs: None

</details>

### PR DESCRIPTION
**Pull request by toobaz**
- [x] closes #23670
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

After "implementing" it (that is, just unifying the method), I'm even more convinced about my [idea](https://github.com/pandas-dev/pandas/issues/23670#issue-380340023)... but @WillAyd and @h-vetinari feel free to consider this just a stimulus to resurrect (and conclude) the discussion ;-)

### CODE REVIEW

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.24.1.rst [hunk 1]
```diff
@@ -69,6 +69,9 @@ Fixed Regressions
 
 - Fixed the warning for implicitly registered matplotlib converters not showing. See :ref:`whatsnew_0211.converters` for more (:issue:`24963`).
 
+**MultiIndex**
+
+- :meth:`Index.to_flat_index` now always returns an :class:`Index` containing tuples, even in the case the original object was already a flat :class:`Index` (:issue:`23670`).
```
[the remark this answers is not in the document]
Comment (toobaz) [reply doc/source/whatsnew/v0.24.1.rst]: This should have been part of #22866, or at least of 0.24.0, when the method was introduced from scratch. Too late for that, but let's at least not expose for long a behavior which is inconsistent. Sure, [this needs discussion](https://github.com/pandas-dev/pandas/issues/23670) ;-)

[the remark this answers is not in the document]
Comment (jreback) [reply doc/source/whatsnew/v0.24.1.rst]: So you mean to say ``Index.to_flat_index`` is not idempotent?

[the remark this answers is not in the document]
Comment (jreback) [reply doc/source/whatsnew/v0.24.1.rst]: why is this a good thing though?

[the remark this answers is not in the document]
Comment (toobaz) [reply doc/source/whatsnew/v0.24.1.rst]: For all the reasons in the [discussion](https://github.com/pandas-dev/pandas/issues/23670), plus one: ``one_level_multi_index.to_flat_index()`` is not idempotent, either.

**Comment by h-vetinari** [comment]
> @toobaz: [...] but @WillAyd and @h-vetinari feel free to consider this just a stimulus to resurrect (and conclude) the discussion ;-)

A bit late to the party here, but I still don't understand the need for flattening MIs. I looked at [the original issue](https://github.com/pandas-dev/pandas/issues/19950) and the usecase there (squashing MI-columns for csv output) could IMO more idiomatically be solved by something like `MI.str.cat(', ')`

Of course, the latter current does not work yet (see also #23679), but I continue to think that tuples should not be usable in indexes. They are already necessary to "talk" to MIs, which makes their use ambiguous, plus it's incredibly hard to get right (and we don't, for large parts of the code base, including basic indexing), not least because it needs so much special casing for interacting with numpy (which treats tuples as arrays by default). See #24688

**Comment by toobaz** [comment]
>  I continue to think that tuples should not be usable in indexes

Will comment on #24688 in few seconds. But this PR assumes that ``to_flat_index`` exists, and tries to fix it. This is a simple PR, worth doing even if ``to_flat_index`` disappears in a couple of releases. Doesn't need all that discussion.

**Review by jreback — CHANGES_REQUESTED** [review]
@toobaz pls in this issue lay out why you think this is a good think. len 1 MultiIndexes, while supported are not very useful. This elevates them to a first class citizen, not a good thing.

**Comment by toobaz** [comment]
> len 1 MultiIndexes, while supported are not very useful. This elevates them to a first class citizen, not a good thing.

No, this PR doesn't change 1 level ``MultiIndex``es in any way.

> @toobaz pls in this issue lay out why you think this is a good think

This is quite clear in the issue: #23670

Anyway, in short, it's really simple:
1) ``Index.to_flat_index()`` doesn't make much sense per se - the index _is_ already flat
2) we support it only for compatibility with ``MultiIndex``
3) ... but in the current form, it is _not_ compatible. The output for ``MultiIndex`` with ``n`` levels is an ``Index`` with tuples of length ``n``
5) So the output for a flat ``Index``, which has 1 level, should be an ``Index`` with tuples of length 1
6) .. which is incidentally the output on a ``MultiIndex`` with 1 level. Which is good, because 1-level ``MultiIndexes`` should behave like flat ``Index``es (I couldn't care less about whether they are popular, encouraged or blamed: as long as they are allowed, we want them to behave consistently)

This is really a simple think, even @WillAyd is [now neutral](https://github.com/pandas-dev/pandas/issues/23670#issuecomment-460010083). Please provide any reason not to merge this, or let's proceed.

**Comment by WillAyd** [comment]
Closing for now - I think it's worth continuing conversation in issue before updating any PRs on this

**Pull request closed without merging**